src/Raven.Server/Documents/ShardedHandlers/ShardedQueriesHandler.cs

```diff
@@ -0,0 +1,843 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Esprima.Ast;
+using Lucene.Net.Search;
+using Lucene.Net.Store;
+using NuGet.Packaging;
+using Raven.Client.Documents.Changes;
+using Raven.Client.Documents.Indexes;
+using Raven.Client.Documents.Queries;
+using Raven.Client.Exceptions.Documents.Indexes;
+using Raven.Server.Documents.Handlers;
+using Raven.Server.Documents.Includes;
+using Raven.Server.Documents.Indexes;
+using Raven.Server.Documents.Indexes.MapReduce.Auto;
+using Raven.Server.Documents.Indexes.MapReduce.Static;
+using Raven.Server.Documents.Indexes.Persistence.Lucene.Documents;
+using Raven.Server.Documents.Patch;
+using Raven.Server.Documents.Queries;
+using Raven.Server.Documents.Queries.AST;
+using Raven.Server.Documents.Queries.Results;
+using Raven.Server.Documents.Queries.Sorting.AlphaNumeric;
+using Raven.Server.Documents.Queries.Timings;
+using Raven.Server.Documents.ShardedHandlers.ShardedCommands;
+using Raven.Server.Documents.Sharding;
+using Raven.Server.Extensions;
+using Raven.Server.Json;
+using Raven.Server.NotificationCenter;
+using Raven.Server.Routing;
+using Raven.Server.ServerWide.Context;
+using Raven.Server.TrafficWatch;
+using Sparrow;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+using Constants = Raven.Client.Constants;
+using HttpMethod = Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpMethod;
+
+namespace Raven.Server.Documents.ShardedHandlers
+{
+    public class ShardedQueriesHandler : ShardedRequestHandler
+    {
+        [RavenShardedAction("/databases/*/queries", "POST")]
+        public Task Post()
+        {
+            return HandleQuery(HttpMethod.Post);
+        }
+
+        [RavenShardedAction("/databases/*/queries", "GET")]
+        public Task Get()
+        {
+            return HandleQuery(HttpMethod.Get);
+        }
+
+        public async Task HandleQuery(HttpMethod httpMethod)
+        {
+            // TODO: Sharding - figure out how to report this
+            var metadataOnly = GetBoolValueQueryString("metadataOnly", required: false) ?? false;
+            using (var tracker = new RequestTimeTracker(HttpContext, Logger, null, "Query"))
+            {
+                try
+                {
+                    //using (var token = CreateTimeLimitedQueryToken())
```

Comment:
consider whether it should be here or only on each shard

Reply:
DevelopmentHelper.ToDo